### DISCUSSION TITLE
Storing Collateral in smart contract for Coin and FA

### DISCUSSION THREAD

**Discussion by ajaythxkur** [aptos-labs/aptos-developer-discussions#407] [Questions] [answered]
I am keeping collateral in object. 
Creating store using constructor_ref,
transferring collateral to store
But this shows the collateral coins in wallet and explorer too
[Image](https://discord.com/channels/945856774056083548/946123778793029683/1282576281732386870)

code:
` let fa = primary_fungible_store::withdraw(giver, fa_metadata, amount);

   let fa_store = fungible_asset::create_store<Metadata>(constructor_ref, fa_metadata); 
   
   fungible_asset::deposit(fa_store, fa);`

**Comment by kshitijc1506** [comment]
I probably need more context here -- do you have a deployment link so I can see the code? 

Also, could you pls be clearer on the issue you're facing? From what I can tell you're trying to store user collateral (APT) in a secondary store, and that is being reflected in your Petra balance? AFAIK Petra only displays balances in primary stores.

**Reply by ajaythxkur** [reply]
Here @kshitjee 
[Explorer](https://explorer.aptoslabs.com/object/0xc1ad5509a57eb43b514e8639967b91fc959bfaab68e606953b07cd1552249e6e/modules/code/nft_lending?network=devnet)
Checkout the `give loan` fn and `Loan struct`

**Reply by kshitijc1506** [reply]
a) Out of curiosity, what's the rationale behind `public(friend) entry` visibility? 

b) Looks like you're withdrawing from the giver's primary store, storing the funds in a secondary store (also owned by the giver), and then storing the store in the `Loan` resource, which is moved to the store object. Looks fine to me. 

The balance on the secondary store will not show up in your Petra wallet. Petra only tracks primary store balances.

**Reply by ajaythxkur** [reply]
a) for writing tests in tests folder
b) when I run the function this actually happens in explorer and wallet. You can try it.

**Reply by kshitijc1506** [reply]
Not sure I follow what the issue is -- if you call `give_loan`, your balance should decrease by `amount`, in the Petra UI. In the explorer, you should be able to see the funds stored in the secondary store.

**Reply by ajaythxkur** [reply]
@kshitjee please follow this images link and check this order [discord images](https://discord.com/channels/945856774056083548/946123778793029683/1283280098841329785)
1> have 2 APT in wallet
2> called `give_loan` function and kept 0.5 apt as collateral
3> still seeing 0.5 apt in petra.
4> When i click send on petra it tries to send the 0.5 collateral apt
is there any way to freeze/unfreeze the FA store??

**Reply by ajaythxkur** [reply]
and is that issue from petra?

**Reply by kshitijc1506** [reply]
thanks for the explanation, that makes sense. there is no transaction on the explorer that indicates `give_loan` txn was successfully executed, is it a different deployment link?

**Reply by ajaythxkur** [reply]
@kshitjee [here](https://explorer.aptoslabs.com/txn/62126230/balanceChange?network=devnet)
see the balance change section too

**Reply by kshitijc1506** [reply]
your petra balance shows up at 2 APT still? interesting... in that case maybe petra does support secondary store balance.

as for why the balance is still shown on the explorer, you're not storing collateral funds into the smart contract. the store you're depositing the 0.5 APT is still owned by the giver:

```rust
let constructor_ref = &object::create_object(address_of(giver));
```

hence the balance change of -0.5 and +0.5.

are you trying to store the collateral in the smart contract? this would probably make more sense

**Reply by ajaythxkur** [reply]
yes i'm trying to store collateral in smart contract. Storing collateral like this is safe right?. I transferred some aptos to another account and its doing nothing to secondary store (convinient). So its the issue from petra about the stores. Thanks for your help. Please let me know about collateral safety. @kshitjee

**Reply by kshitijc1506** [reply]
to store funds in the smart contract you can create an object in init_module, retain the signer using extend_ref, and use that signer to create the secondary store in `give_loan`. 

this is one of way saying your smart contract is the owner of the collateral funds, since the deployer signer is the code object, and it owns the object that creates the secondary stores. 

to withdraw from a store, you will need to use the extend_ref again to generate a signer.

**Reply by kshitijc1506** [reply]
so code object owns "creation/signer" object 

creation object uses signer and owns secondary store object 

this is safe since there's no private key involved, and only the extend_ref can be used to withdraw funds from a collateral vault

**Reply by ajaythxkur** [reply]
thanks @kshitjee this makes more sense, i'll update the code.
one more thing, is it possible to do it with coin. i'll use both coins and fa as collateral. but i have one question in mind, when running repay function, how can i check the correct type of coin being returned?

**Reply by ajaythxkur** [reply]
and how can multiple coins be stored in contract?

**Reply by kshitijc1506** [reply, marked as answer]
to support coin and fa you have two options:

(1) create a coin to fungible asset wrapper, here's an [example](https://github.com/aptos-labs/aptos-core/blob/d45289431871d8c5c85d91a8b126a5981089a7b9/aptos-move/move-examples/swap/sources/coin_wrapper.move)

(2) create separate entry functions for coin and fa, here's a rewards [contract](https://explorer.aptoslabs.com/account/0x02bf331b3010eccd0cc255e7338220d9c5c5aff1c363e230c80a7a08da608860/modules/code/ae_rewards?network=mainnet) that stores both fungible assets and coins to later distribute as airdrops. 

although, ^this doesn't store both coin and fa inside an object, it stores them in the admin account, similar to what you're doing. but that made sense for the requirement. 

also, here's aave v3 implemented in move for [reference](https://github.com/aave/aptos-aave-v3/tree/78738daa269db5445bc8697e62a822c1aa6425ff) :)

**Reply by kshitijc1506** [reply]
oh option (3) is to burn [coin and mint fa](https://github.com/aptos-labs/aptos-core/blob/33a72814b63268db96098c96aea2ca5cfc99dfee/aptos-move/framework/aptos-framework/sources/coin.move#L381), but i'd be careful to do this on mainnet, since it's probably smarter to wait for the ecosystem to completely migrate to fungible asset first.
